File: bindings/python/tests/test_binding.py
import shutil
import tempfile
import unittest
import pomaidb

class TestPomaiDBBinding(unittest.TestCase):
    def setUp(self):
        self.test_dir = tempfile.mkdtemp(prefix="pomai_py_test_")

    def tearDown(self):
        shutil.rmtree(self.test_dir, ignore_errors=True)

    def test_database_class_api(self):
        db = pomaidb.Database.open(self.test_dir, dim=4)
        self.assertIsNotNone(db)

        # Put vector
        db.put(1, [1.0, 0.0, 0.0, 0.0])
        db.create_membrane("mem1", dim=4)
        db.open_membrane("mem1")
        db.put(2, [0.0, 1.0, 0.0, 0.0], membrane="mem1", payload=b"payload_val", timestamp=42)

        # Exists
        self.assertTrue(db.exists(1))
        self.assertTrue(db.exists(2, membrane="mem1"))

        # Get
        v1 = db.get(1)
        self.assertEqual(len(v1), 4)
        self.assertAlmostEqual(v1[0], 1.0, places=3)

        rec2 = db.get(2, membrane="mem1", with_metadata=True)
        self.assertEqual(rec2.payload, b"payload_val")
        self.assertEqual(rec2.timestamp, 42)

        # Search
        res = db.search([1.0, 0.0, 0.0, 0.0], topk=2)
        self.assertTrue(len(res) >= 1)
        self.assertEqual(res[0].id, 1)

        # Stats
        stats = db.get_stats()
        self.assertIn("version", stats)

        db.flush()
        db.close()

    def test_functional_c_api(self):
        dir2 = tempfile.mkdtemp(prefix="pomai_py_c_")
        try:
            h = pomaidb.open_db(dir2, dim=4)
            pomaidb.put(h, 10, [0.5, 0.5, 0.5, 0.5])
            self.assertTrue(pomaidb.exists(h, 10))
            v = pomaidb.get(h, 10)
            self.assertEqual(v["id"], 10)
            pomaidb.flush(h)
            pomaidb.close(h)
        finally:
            shutil.rmtree(dir2, ignore_errors=True)

if __name__ == "__main__":
    unittest.main()

File: bindings/rust/tests/test_binding.rs
use pomaidb::{Database, MetricType, Options, PutOptions, QuantType};
use std::fs;
use std::time::{SystemTime, UNIX_EPOCH};

#[test]
fn test_basic_crud_and_search() {
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis();
    let test_dir = format!("tmp_rust_test_{}", now);

    let opts = Options::new(&test_dir, 4)
        .metric(MetricType::L2)
        .quant_type(QuantType::None);

    let mut db = Database::open(&opts).expect("Failed to open db");

    // Put item
    db.put(1, &[1.0, 0.0, 0.0, 0.0]).expect("put failed");
    assert!(db.exists(1).expect("exists check failed"));

    // Multi-membrane
    db.create_membrane("test_mem", 4, 1).expect("create_membrane failed");
    db.open_membrane("test_mem").expect("open_membrane failed");

    db.put_with_options(
        2,
        &[0.0, 1.0, 0.0, 0.0],
        &PutOptions {
            membrane: Some("test_mem"),
            timestamp: 1234567,
            payload: Some(b"hello rust"),
        },
    ).expect("put_with_options failed");

    assert!(db.exists_membrane(Some("test_mem"), 2).expect("exists_membrane failed"));

    let rec = db.get_membrane(Some("test_mem"), 2).expect("get_membrane failed").expect("record 2 missing");
    assert_eq!(rec.id, 2);
    assert_eq!(rec.timestamp, 1234567);
    assert_eq!(rec.payload.as_deref(), Some(b"hello rust".as_slice()));

    // Search
    let hits = db.search(&[1.0, 0.0, 0.0, 0.0], 2).expect("search failed");
    assert!(!hits.is_empty());
    assert_eq!(hits[0].id, 1);

    // Stats
    let stats = db.get_stats().expect("get_stats failed");
    assert!(stats.get("version").is_some());

    db.flush().expect("flush failed");
    db.close().expect("close failed");

    let _ = fs::remove_dir_all(&test_dir);
}